### DISCUSSION TITLE
No update check on startup?

### DISCUSSION THREAD

**Discussion by Devidian** [mbround18/valheim-docker#197] [Help!]
I have set the restart & update to 05:15 on my server because this is the time where nobody will play. So the update came on 09:00 and of course the server is version mismatching. Also odin panicked somehow - so i restarted the docker container but it is still on the previous version. This behavior is not correct in my opinion. Now i have to delete binary again to redownload on container restart.

What about changing the auto update to "Update on startup/restart" and only have a "auto restart" on cron?

AUTO_UPDATE = 1  => check for updates on startup
+AUTO_RESTART = 1 => enable auto-restart
AUTO_UPDATE_SCHEDULE => AUTO_RESTART_SCHEDULE

**Comment by CosmicHorrorDev** [comment]
So just a couple things note

- You typically should _not_ have to delete the server files to update
- If `odin` panicked then it's good to report why here, so that it can get fixed
- Updating on startup is in the works. I'd be in favor of `UPDATE_ON_STARTUP` as the name since `AUTO_RESTART` seems misleading

It was a pretty painful day update-wise since there was an update yesterday and another today. I appreciate hearing the feedback since it lets me know people's pain points.

There is also work going on to check for updates without having to shutdown the server which should open the door for more frequent update checking.

**Reply by CosmicHorrorDev** [reply]
And to attempt an update you can run

```text
$ docker exec -it --user steam $CONTAINER_NAME bash ../scripts/auto_update.sh
```

Let me know if you run into any difficulties here!

**Reply by Devidian** [reply]
> * Updating on startup is in the works. I'd be in favor of `UPDATE_ON_STARTUP` as the name since `AUTO_RESTART` seems misleading

Oh i meant AUTO_RESTART as "its only doing a scheduled restart - nothing more" Once we have UPDATE_ON_STARTUP we dont need AUTO_UPDATE scheduled i think.

**Comment by CosmicHorrorDev** [comment]
Just to give some updates from recent changes. Thanks to feedback; checking for updates on startup was added already instead of waiting for #177 to be finished (in #202 and #204)!

Checking for updates on startup is exposed through the `UPDATE_ON_STARTUP` env var which is **on** by default! Example logs from a server checking for updates when launching:

```text
[Valheim][steam]: Running Install...
[Valheim][steam]: Attempting to update before launching the server!    <-- This line here is the indicator
[Valheim][steam]: ###########################################################################
[Valheim][steam]: Valheim Server - Sat Feb 27 02:42:54 PST 2021
[Valheim][steam]: Starting auto update...
```

So if you want the server to check for updates when starting you just need to pull the latest image :D

**Reply by Devidian** [reply]
cool nice and by the way as i see it just now

```log
Connecting anonymously to Steam Public...Logged in OK
Waiting for user info...OK
Success! App '896660' already up to date.
[ODIN][INFO]  - Successfully installed Valheim!
[Valheim][steam]: Starting server...
[ODIN][INFO]  - Run with DEBUG_MODE as 1 if you think there is an issue with Odin
[ODIN][INFO]  - Setting up start scripts...
thread 'main' panicked at 'Please initialize odin with `odin configure`. See `odin configure --help`', src/files/config.rs:22:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

==> /home/steam/valheim/logs/auto-backup.out <==
```

**Reply by CosmicHorrorDev** [reply]
Thanks for the error log! It does look like adding in `auto_update.sh` also added in a bug that only shows up in certain circumstances. I'll open a PR to fix it, but in the meantime you can run

```bash
docker exec --user steam -w /home/steam/valheim $CONTAINER_NAME odin configure
```

to fix your issue!

**Reply by Devidian** [reply]
> ```shell
> docker exec --user steam -w /home/steam/valheim $CONTAINER_NAME odin configure
> ```

I've tried it but every day i read the same error again.

```log
==> /home/steam/valheim/logs/auto-update.out <==
[ODIN][INFO]  - Server has been shutdown successfully!
[Valheim][steam]: Starting auto backup process...
[ODIN][INFO]  - Run with DEBUG_MODE as 1 if you think there is an issue with Odin
[Valheim][steam]: Backup process complete! Created 20210302-050016-pre-update-backup.tar.gz
[Valheim][steam]: Installing Updates...
[ODIN][INFO]  - Run with DEBUG_MODE as 1 if you think there is an issue with Odin
[ODIN][INFO]  - Installing 896660 to /home/steam/valheim
[ODIN][ERROR] - Failed to find steamcmd in path
[ODIN][ERROR] - Checking for script under steam user.
[ODIN][INFO]  - Executing: /home/steam/steamcmd/steamcmd.sh .....
[ODIN][INFO]  - Using steamcmd script at /home/steam/steamcmd/steamcmd.sh
WARNING: setlocale('en_US.UTF-8') failed, using locale: 'C'. International characters may not work.
Redirecting stderr to '/home/steam/Steam/logs/stderr.txt'
[  0%] Checking for available updates...
[----] Verifying installation...
Steam Console Client (c) Valve Corporation
-- type 'quit' to exit --
Loading Steam API...OK.

Connecting anonymously to Steam Public...Logged in OK
Waiting for user info...OK
Success! App '896660' already up to date.
[ODIN][INFO]  - Successfully installed Valheim!
[Valheim][steam]: Starting server...
[ODIN][INFO]  - Run with DEBUG_MODE as 1 if you think there is an issue with Odin
[ODIN][INFO]  - Setting up start scripts...
thread 'main' panicked at 'Please initialize odin with `odin configure`. See `odin configure --help`', src/files/config.rs:22:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

The configure task seems to create the config file, but this is already existing

```log
insgesamt 65172
-rw-r--r-- 1 1000 1000      151 Mär  1 08:28  config-01.json
-rw-r--r-- 1 1000 1000      164 Mär  1 08:25  config-02.json
drwxr-xr-x 2 1000 1000     4096 Feb 26 13:34  linux64
-rwxr-xr-x 1 1000 1000     4728 Feb 26 13:34  LinuxPlayer_s.debug
drwxr-xr-x 2 root root     4096 Feb 26 13:33  logs
-rwxr-xr-x 1 1000 1000        2 Feb 26 13:34  server_exit.drp
-rwxr-xr-x 1 1000 1000      575 Feb 26 13:34  start_server.sh
-rwxr-xr-x 1 1000 1000       34 Feb 26 13:34  start_server_xterm.sh
-rwxr-xr-x 1 1000 1000        7 Feb 26 13:34  steam_appid.txt
drwxr-xr-x 5 1000 1000     4096 Feb 26 13:34  steamapps
-rwxr-xr-x 1 1000 1000 30012405 Mär  1 08:28  steamclient.so
-rwxr-xr-x 1 1000 1000  7041000 Feb 26 13:34  UnityPlayer_s.debug
-rwxr-xr-x 1 1000 1000 29484584 Feb 26 13:34  UnityPlayer.so
-rwxr-xr-x 1 1000 1000   131994 Feb 26 13:34 'Valheim Dedicated Server Manual.pdf'
drwxr-xr-x 6 1000 1000     4096 Feb 26 13:34  valheim_server_Data
-rwxr-xr-x 1 1000 1000     6344 Feb 26 13:34  valheim_server.x86_64
```

it is the config-01.json

Also i noticed that starting more than one server instance at once seems to crash the older ones - might be caused by the update on startup - will test it by only allowing one instance to update.

**Reply by Devidian** [reply]
It might be helpful if the container would exit in that case - because docker would restart it with restart set - in compose file for example.

**Reply by mbround18** [reply]
Since you are running with multiple servers can you set the flag below? 

```sh
UPDATE_ON_STARTUP=0
```

**Reply by Devidian** [reply]
i did but it crashed anyways. It was running fine before but it seems like its not the issue. What files can i provide to help?

[server01.log](https://github.com/mbround18/valheim-docker/files/6066591/server01.log)
[server01.err.log](https://github.com/mbround18/valheim-docker/files/6066594/server01.err.log)
[server02.log](https://github.com/mbround18/valheim-docker/files/6066593/server02.log)
[server02.err.log](https://github.com/mbround18/valheim-docker/files/6066592/server02.err.log)

I've started the second server with 5 minutes delay

**Reply by mbround18** [reply]
Can you send me you docker compose? minus the passwords and all

**Reply by mbround18** [reply]
Do you use the cron job update feature? or do you prefer to manually update your servers?

**Reply by Devidian** [reply]
i would like to use the cron, but this seems to get the server into the panicked state
[docker-compose-gs01.yml.log](https://github.com/mbround18/valheim-docker/files/6067381/docker-compose-gs01.yml.log)
[docker-compose-gs02.yml.log](https://github.com/mbround18/valheim-docker/files/6067382/docker-compose-gs02.yml.log)

**Reply by mbround18** [reply]
> i would like to use the cron, but this seems to get the server into the panicked state
> [docker-compose-gs01.yml.log](https://github.com/mbround18/valheim-docker/files/6067381/docker-compose-gs01.yml.log)
> [docker-compose-gs02.yml.log](https://github.com/mbround18/valheim-docker/files/6067382/docker-compose-gs02.yml.log)

Can you pull the latest update? I just published a version which sets a few environment variables for Odin to consume.

**Reply by Devidian** [reply]
[server01.log](https://github.com/mbround18/valheim-docker/files/6067604/server01.log)
[server02.log](https://github.com/mbround18/valheim-docker/files/6067606/server02.log)
[server02.err.log](https://github.com/mbround18/valheim-docker/files/6067605/server02.err.log)
[server01.err.log](https://github.com/mbround18/valheim-docker/files/6067607/server01.err.log)

The first server crashed again when the second came up, but this time no message of failure in the log, just in the err.log

**Reply by Devidian** [reply]
after the recent game update both seem to run again, i will keep an eye on that. At least tomorrow after the update schedule i should know more.

**Reply by Devidian** [reply]
Well seems to run so far, at least today both instances are still running. It just looks like none has restarted for scheduled update checks.
I got discord messages of both have stopped on 5:00 and 5:15 - but no `Start` message - so i thought both were down - but they are up and running 
![image](https://user-images.githubusercontent.com/10413872/109759778-e38bfd00-7bed-11eb-8157-054e8f43aa55.png)
CPU time just doesnt look like they ever have restarted 2 hours ago :D
